src/concurrency/thread.rs

```diff
@@ -184,14 +186,17 @@ impl<'mir, 'tcx> Thread<'mir, 'tcx> {
 #[derive(Debug)]
 pub enum Time {
     Monotonic(Instant),
+    Virtual { nanoseconds: u64 },
     RealTime(SystemTime),
 }
 
 impl Time {
     /// How long do we have to wait from now until the specified time?
-    fn get_wait_time(&self) -> Duration {
+    fn get_wait_time(&self, clock: &Clock) -> Duration {
         match self {
             Time::Monotonic(instant) => instant.saturating_duration_since(Instant::now()),
+            Time::Virtual { nanoseconds } =>
+                Duration::from_nanos(nanoseconds - clock.assert_virtual().load(Ordering::Relaxed)),
```

Comment:
All of the pattern matching between `Monotonic` and `Virtual` should be isolated inside `time.rs`. I think this can be achieved by having that file just export a type `Instant` and using that in the rest of Miri.

I think you can make that type just be a simple `u64` (nanoseconds), interpreted relative to the time_anchor for 'real' monotone time. But you can also make it an `enum` that is either a `std::time::Instant` or a `u64`, I don't have a strong preference either way. But it should be private to `shims/time.rs`.